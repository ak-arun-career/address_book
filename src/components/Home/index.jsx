import { nanoid } from 'nanoid';
import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import ContactActions from '../../redux/Contacts';
import { Contacts } from '../Contacts';
import ContactForm from '../Contacts/ContactForm';
import Header from '../Header';

export const Home = () => {
    const [showContactForm, setShowContactForm] = useState(false);
    const [formAction, setFormAction] = useState(null);
    const dispatch = useDispatch();

    /** Function to display the contact form
     * In case a contact is edited, the form comes pre-filled with the contact details
     */
    const showForm = (data, action) => {
        if (action === 'edit') {
            dispatch(ContactActions.actions.addToEdit(data))
        }
        setShowContactForm(true);
        setFormAction(action);
    };

    /** Function to generate a unique id to be used while adding a contact */
    const generateUniqueId = (length = 4) => {
        return nanoid(length);
    };

    /** Submit form */
    const onSubmitContactForm = data => {
        closeForm();

        const contactData = {
            "id": (data.id || generateUniqueId()),
            "name": {
                "first": data.firstName,
                "last": data.lastName
            },
            "email": data.email,
            "phone": data.phone,
            "status": data.status
        };

        /* Dispatch respective action based on formAction */
        (formAction === 'edit') ? dispatch(ContactActions.actions.editContact(contactData)) : dispatch(ContactActions.actions.addContact(contactData));
    };

    /** Function to close the contact form */
    const closeForm = () => {
        setShowContactForm(false);
        dispatch(ContactActions.actions.clearAfterEdit());
    };

    return (
        <>
            <Header showForm={() => showForm({}, 'add')} />

            {showContactForm &&
                <ContactForm
                    submitForm={(data) => onSubmitContactForm(data)}
                    cancelForm={() => closeForm}
                />
            }

            <Contacts
                editContact={(data) => showForm(data, 'edit')}
            />
        </>
    );
};