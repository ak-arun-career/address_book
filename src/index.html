<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Address Book</title>
    <style>
        * {
            margin: 0;
            padding: 0;
        }
        body {
            font-family: Roboto, 'Helvetica Neue', sans-serif;
            background: lavender;
        }
    </style>
</head>

<body>
    <div id="root"></div>
</body>

</html>