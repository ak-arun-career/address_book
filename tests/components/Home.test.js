import React from 'react';
import { shallow } from 'enzyme';
import { App } from '../../src/App';
import { Home } from 'Components/Home';

describe('App', () => {
    it("renders without error", () => {
        const page1 = shallow(<App />);
        expect(page1.length).toEqual(1);
    });

    it("should render the child component correctly", () => {
        const child = "Contact";
        const page2 = shallow(<App>{child}</App>);
        expect(page2).toMatchSnapshot();
    });

    it('should render a <Home /> component', () => {
        const page3 = shallow(<App />);
        expect(page3.find(Home)).toHaveLength(1);
    })
})