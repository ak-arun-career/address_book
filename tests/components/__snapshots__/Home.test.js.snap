// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render the child component correctly 1`] = `ShallowWrapper {}`;
